Task Authoring and Monitoring Improvements

shirishgoyal opened the pull request:
**CSV**
a) CSV data sources can be added anywhere inside content of label or value of a component
![image](https://cloud.githubusercontent.com/assets/315012/10560358/a0e437fc-7526-11e5-9608-ae2bb554b05b.png)

b) CSV items can be copied to clipboard on clicking and can be pasted to avoid typing

**Design Canvas**
a) Components can be added without drag/drop
![image](https://cloud.githubusercontent.com/assets/315012/10560446/4705fc90-7529-11e5-842c-a0b24c16c286.png)

b) Same component can be used for display or editing inspired from Google Forms
![image](https://cloud.githubusercontent.com/assets/315012/10560377/4a65499c-7527-11e5-8a80-fdc332dfd164.png)

c) Each component (except Instructions) has a label and options/placeholder. Label can be directly associated to Answer provided by worker

d) Instructions allow rich text/html content for better formatting
![image](https://cloud.githubusercontent.com/assets/315012/10560385/904ce62c-7527-11e5-840b-c5635f7e07e4.png)

e) Updated Preview for above changes to template compiler and CSV

f) Consolidated Labeled Input component and Text Input into single component
![image](https://cloud.githubusercontent.com/assets/315012/10560441/33e27f9e-7529-11e5-930b-27338eeed4f7.png)

**Task**
a) Updated task view for above changes to template compiler and CSV

**Task Monitoring**
a) Replaced display items with Label values from Components in Task

shirishgoyal commented:
WIP: 
a) Checkbox - handle worker response
b) Handle html entities in CSV headers

dmorina commented on bower.json:
```diff
@@ -6,20 +6,23 @@
-    "underscore": "1.8.3"
+    "underscore": "1.8.3",
+    "ckeditor": "~4.5.4",
+    "angular-ckeditor": "~1.0.0"
```
ckeditor breaks design, if we are gonna have a rich text editor (which we dont need atm) we are gonna implement it using MD

dmorina commented:
drop a) and b)
-bring back autocomplete for data source
-drop d) rich text
-add item should stay at the bottom not covering the existing items
-on hover on an item it should show at least Edit, Duplicate, Delete quick actions
-on edit mode use all space for properties

others might follow after iteration

aginzberg commented:
appearing a bit funky on my computer. I did a bower install so im not sure what is up. is it not supported on chrome? and it doesnt quite look right on firefox. do i need to do something else? see the screenshots below.

Chrome:
<img width="983" alt="screen shot 2015-11-04 at 11 04 08 am" src="https://cloud.githubusercontent.com/assets/4563883/10948739/f7acc214-82e3-11e5-974e-b9da98d9f52e.png">

Firefox:
<img width="981" alt="screen shot 2015-11-04 at 11 04 36 am" src="https://cloud.githubusercontent.com/assets/4563883/10948752/065bbd9c-82e4-11e5-9dfe-dc766417cd85.png">

aginzberg commented:
More suggested changes
1. Be able to navigate by clicking inside of a template item.
2. Text like "Add question here" should be a placeholder rather than actual text in the label.
3. Click or double-click to edit template item.
4. Unable to delete first element.
5. Do not allow Text fields, radio buttons, etc be editable when not in edit mode.

shirishgoyal commented:
Task Monitoring Interface
1. Merged Input with Answer
2. Task row can expand to show task detail
3. Rating merged with task row
4. Feedback merged into task detail
5. Bookmark improved
6. Rating Instructions merged and moved to top

![image](https://cloud.githubusercontent.com/assets/315012/10997809/3ef616b0-84b6-11e5-8dff-41b02ca518ef.png)

shirishgoyal commented:
Updated Task Design Interface

![image](https://cloud.githubusercontent.com/assets/315012/10999074/061c6c90-84c0-11e5-8394-e4c56f0a6eee.png)

dmorina commented on crowdsourcing/models.py:
```diff
@@ -274,7 +274,7 @@ class TemplateItem(models.Model):
     data_source = models.CharField(max_length=256, null=True)
     layout = models.CharField(max_length=16, default='column')
     type = models.CharField(max_length=16)
-    sub_type = models.CharField(max_length=16)
```
we want to keep subtype, will be useful for example for groups of items e.g type: input sub_type: number, but additional changes need to be made so it's fine for now

shirishgoyal replied:
I would lean towards adding new component - number than adding specific details to input component. Number can be integer, decimal, range etc. so better to keep it separate.

dmorina replied:
we map 1:1 to html so when I say I want a number input field I cannot specify the datatype other than number, or email or text otherwise the logic becomes more complicated unnecessarily but this comes later no worries

dmorina commented:
tried submitting a project 
I get {"modules":[{"template":[{"template_items":[{"values":["This field may not be blank."]},{"values":["This
 field may not be blank."]},{},{}]}]}]}, template looks as follows

dmorina commented:
<img width="820" alt="screen shot 2015-11-06 at 11 10 27 am" src="https://cloud.githubusercontent.com/assets/2858281/11006168/1ad11c74-8477-11e5-9567-3159c5f1ca95.png">

shirishgoyal commented:
Modified the instructions component to use label instead of values so that we can use values solely for passing options or input values which need worker's input. Instructions is merely a display component so makes it easy to filter at other places.

Added a default text for values for Instructions.

dmorina commented:
@shirishgoyal lgtm, could you just merge carefully the tabindex changes and then we are good to go

shirishgoyal commented:
Updated tab support in task display interface

For radio-buttons in material design, tab support for options added instead of default radio-group. 
Instead of spacebar, arrow keys work for selection of choice. Please see comments here: https://github.com/angular/material/issues/3954

dmorina commented:
lgtm

aginzberg commented:
for rating on the task monitoring, I think we want to leave it as it was because we aren't rating each individual piece of work done but the worker holistically. i will raise another pr for refining this to only show a few (chosen intelligently) and support search. also we can move the blue box at the top back over to the side rating thing.

shirishgoyal commented:
@aginzberg: the rating logic remains the same. It happens at individual level. It is only shown once per task row.

dmorina commented:
The thing is that you are not rating tasks and we dont wanna show more than 5 workers to rate, that's why it used to be separate, these two things shouldn't be mixed, main purpose of this page is monitoring progress and results, in addition requesters have the option to rate their top 5 workers or others too

dmorina commented on static/js/task/controllers/task.controller.js:
```diff
@@ -58,16 +58,7 @@
-                        }
-                        angular.extend(obj, {tabIndex: tabIndex});
-                    });
+
```
reason for reverting this one?

shirishgoyal replied:
We only add tabIndex to components which need to be excluded from tabs. The rest of the components follow the default indexing which works as intended with above code.

dmorina commented:
@shirishgoyal tabindex changes have been lost somewhere?! was that intentional? if not can you bring them back
